Replace go-bindata with //go:embed

palnabarun opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

With kubernetes/kubernetes using Go 1.16 (ref: https://github.com/kubernetes/kubernetes/pull/98572), we should be exploring at better ways of packaging/bundling binary data.

The package that we currently use for embedding binary data (go-bindata) is largely unmaintained and we have had to update references to the package (ref: https://github.com/kubernetes/kubernetes/issues/96169). Go 1.16 has built-in support for embedding files in the compiled binaries.

This pull request explores removing the dependency on go-bindata. Currently, we use go-bindata for packaging the following:
- kubectl translations data
- Conformance test fixtures

##### Tasks

- [x] Migrate kubectl translation data
- [x] Migrate conformance text fixtures
- [x] Remove go-bindata related artifacts

#### Which issue(s) this PR fixes:

Fixes #99150

#### Special notes for your reviewer:

##### kubectl translations
 
Go 1.16's embed directive doesn't allow embedding files from parent directories. Hence, moving the translations data to inside the i18n package.

Logically speaking as well, kubectl related artifacts should be inside the kubectl package. I think historically the files were at the root of k/k because of the generate script.

Comparing compilation times and binary size for kubectl:

```
(on this branch)
❯ time make WHAT=./cmd/kubectl
...
make WHAT=./cmd/kubectl  336.24s user 31.96s system 315% cpu 1:56.56 total
❯ ls -l _output/local/bin/linux/amd64/kubectl
-rwxr-xr-x 1 nabarun nabarun 46346240 Jun  4 10:02 _output/local/bin/linux/amd64/kubectl

(on master)
❯ time make WHAT=./cmd/kubectl
...
make WHAT=./cmd/kubectl  348.77s user 39.83s system 318% cpu 2:01.97 total
❯ ls -l _output/local/bin/linux/amd64/kubectl
-rwxr-xr-x 1 nabarun nabarun 46366720 Jun  4 07:23 _output/local/bin/linux/amd64/kubectl
```

##### Conformance test fixtures

The conformance test fixtures are spread over several packages and hence not as trivial to embed. 

	"test/conformance/testdata/..."
	"test/e2e/testing-manifests/..."
	"test/e2e_node/testing-manifests/..."
	"test/images/..."
	"test/fixtures/..."

Inside each of the above directories, there is a function which needs to be called whenever those fixtures/testdata need to be used. 

##### References

https://golang.org/pkg/embed/
https://go.googlesource.com/proposal/+/master/design/draft-embed.md

#### Does this PR introduce a user-facing change?

```release-note
Replace go-bindata with //go:embed 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

/assign
/priority important-soon

/sig architecture
/cc @dims

/sig cli
/area kubectl
/cc @soltysh

palnabarun committed:
kubectl: move translations to i18n package to kubectl staging directory

Go 1.16's embed directive doesn't allow embeding files from parent
directories. Hence, moving the translations data to inside the i18n package.

Logically speaking as well, kubectl related artifacts should be inside
the kubectl package.

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
kubectl: use embedded translations instead of generated bindata

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
hack: remove bindata generation logic for translations

also, remove generated bindata

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
vendor: update manifest

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

soltysh commented:
@palnabarun wow this looks great! the translation should've been moved along with the rest of kubectl last release, thanks!

palnabarun commented:
/area code-organization

dims commented:
@palnabarun can you please see if you can prioritize this? would be good to land this early in 1.22

palnabarun committed:
e2e-framework: add an EmbeddedFileSource to sources

The data structure would wrap an embedded filesystem andthe root
directory relative to which the embedded filesystem is constructed.

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
test: setup embedded file sources for manifests

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
generated: remove usage of go-bindata

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
e2e-node: move to embedded test manifests

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun committed:
dependencies: remove go-bindata

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

palnabarun commented:
/test pull-kubernetes-conformance-image-test

dims commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
Woooot!!! All tests passed.

I will clean up the commit history on this and add test cases for the test files reading logic. I noticed that tests weren't present earlier either. That will mark the changes reviewable.

BenTheElder commented:
- How does this affect compile times, now that instead of statically generated files we include sources using the go tool at build time?
- How does this affect library users (thinking kubectl)? At minimum they must need new enough go to support this? Does the go module server / dependency management handle this correctly? (like the source export zips would now need to have the source files instead of the generated go files)?

palnabarun commented:
> How does this affect compile times, now that instead of statically generated files we include sources using the go tool at build time?

I tried on just one platform (linux/amd64) and the compile time was similar in both cases. It was a bit faster with the changes here, but can be random behavior as well. I am yet to test the same on other platforms. I need to dig deeper into our release tooling to see how we compile our user-facing binaries.

The binary sizes are also similar, with the binary without embedded files a bit smaller.
```
(on this branch)
❯ time make WHAT=./cmd/kubectl
...
make WHAT=./cmd/kubectl  336.24s user 31.96s system 315% cpu 1:56.56 total
❯ ls -l _output/local/bin/linux/amd64/kubectl
-rwxr-xr-x 1 nabarun nabarun 46346240 Jun  4 10:02 _output/local/bin/linux/amd64/kubectl
```

```
(on master)
❯ time make WHAT=./cmd/kubectl
...
make WHAT=./cmd/kubectl  348.77s user 39.83s system 318% cpu 2:01.97 total
❯ ls -l _output/local/bin/linux/amd64/kubectl
-rwxr-xr-x 1 nabarun nabarun 46366720 Jun  4 07:23 _output/local/bin/linux/amd64/kubectl
```

The full log is [here](https://gist.github.com/palnabarun/35dcb05f415eccb527a89844bdbd7a07)

> How does this affect library users (thinking kubectl)? At minimum they must need new enough go to support this?

If the users are using the compiled binary that we publish, I don't see any change in behavior. If they are pulling in source code and building things (like kubectl) on their own, they must have Go 1.16, which is a sane ask since k/k depends on Go 1.16.

The same goes for when someone is importing `k8s.io/kubectl`, the package is importable in projects using a Go version supporting the embed package. Otherwise, they would get an error like the following:

```
❯ go1.15 run main2.go
../../kubernetes/kubectl/pkg/util/i18n/i18n.go:22:2: package embed is not in GOROOT (/home/nabarun/sdk/go1.15/src/embed)
```

>  Does the go module server / dependency management handle this correctly? (like the source export zips would now need to have the source files instead of the generated go files)?

I have to explore more this. If the module server treats go source files and other filetypes differently, there _maybe_ an issue.

palnabarun commented:
> (like the source export zips would now need to have the source files instead of the generated go files)?

@BenTheElder --  do you have any leads on how I can test if anything changed wrt source export zips locally?

I am reading https://golang.org/ref/mod#module-proxy and https://golang.org/ref/mod#zip-files to get some thoughts.

BenTheElder commented:
> @BenTheElder -- do you have any leads on how I can test if anything changed wrt source export zips locally?

I'm not sure. I don't know what all the zips trim if anything either, but the history of go dependency management involves a lot of "trimming non-source files", so just a sanity check 😅  I _imagine_ go handles this as expected, but a quick glance didn't uncover anything, and it stuck out to me as a potential difference versus generated checked-in source files.

dims commented:
/milestone v1.22

palnabarun commented:
/retitle Replace go-bindata with //go:embed

palnabarun commented on staging/src/k8s.io/kubectl/pkg/util/i18n/translations/README.md:
```diff
@@ -42,6 +42,10 @@ We use the English translation as the `msgid`.
 
 ## Regenerating the bindata file
 
+> Note: Regeneration of bindata is no more necessary for Kubernetes 1.22+ as
+> the translations are now embedded into the binary at compile time.
+> See: https://github.com/kubernetes/kubernetes/pull/99829
+
```
@soltysh -- how about this note?

I couldn't think of convincing scenarios where translations to released Kubernetes versions will be updated, but still felt to keep the block and add the note.

palnabarun commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -89,12 +94,13 @@ func registerNodeFlags(flags *flag.FlagSet) {
-		AssetNames: generated.AssetNames,
-	})
-
+	// Enable embedded FS file lookup as fallback
+	// TODO: This needs to be cleaned up later to include only file sources which are necessary.
+	e2etestfiles.AddFileSource(e2etestingmanifests.GetE2ETestingManifestsFS())
+	e2etestfiles.AddFileSource(e2enodetestingmanifests.GetE2ENodeTestingManifestsFS())
+	e2etestfiles.AddFileSource(testimages.GetTestImagesFS())
+	e2etestfiles.AddFileSource(testfixtures.GetTestFixturesFS())
+	e2etestfiles.AddFileSource(conformancetestdata.GetConformanceTestdataFS())
```
@kubernetes/sig-node-pr-reviews -- does the e2e_node test suite require test fixtures from all the sources?

I noticed the usage of "test/e2e_node/testing-manifests", asking to be sure about all other test data usages and for backward compatibility.

palnabarun commented:
> > @BenTheElder -- do you have any leads on how I can test if anything changed wrt source export zips locally?
> 
> I'm not sure. I don't know what all the zips trim if anything either, but the history of go dependency management involves a lot of "trimming non-source files", so just a sanity check sweat_smile I _imagine_ go handles this as expected, but a quick glance didn't uncover anything, and it stuck out to me as a potential difference versus generated checked-in source files.

I tried this by pushing [a tag](https://github.com/palnabarun/kubectl/tree/v0.99.0-nabarun.1) in my fork with the translation files.

A `go mod download` of that version of the source code had all the non-source files as well in the go mod cache. I feel this should give a good signal that the non-source files (translations, test fixtures etc.) will be handled correctly.

dims commented:
@palnabarun looks like one thing left

```
Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/test/e2e/framework/testfiles/testfiles_test.go
```

dims commented:
@palnabarun can you please add binary sizes before and with this PR?

palnabarun commented:
> @palnabarun looks like one thing left
> 
> ```
> Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/test/e2e/framework/testfiles/testfiles_test.go
> ```

Done. Needed a new line. :lolsob:

palnabarun commented:
/test pull-kubernetes-unit

palnabarun commented:
/test pull-kubernetes-conformance-image-test

liggitt commented on staging/src/k8s.io/kubectl/pkg/util/i18n/translations/README.md:
```diff
@@ -42,6 +42,10 @@ We use the English translation as the `msgid`.
+
 With the `mo` files up to date, you can now convert the generated files
 into code using `go-bindata` command which can be installed with:
 
```
does the moved `extract.py` script still work in its new location?

palnabarun replied:
Good point. I didn't verify that. Will check the same.

palnabarun replied:
Checked. It does work. The script takes in a file path as an argument. Changing the location of the script itself doesn't change its working.

The documentation just needs to reflect the latest location.

liggitt commented on test/conformance/testdata/embed.go:
```diff
@@ -0,0 +1,33 @@
+	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed *
```
should this be *.yaml? otherwise it embeds itself, right?

palnabarun replied:
Yes, it does embed itself. I thought about it earlier and wondered it was a small file, so wouldn't do much harm.

But, after thinking once more, let's just specifically say what we want to embed.

liggitt commented on test/e2e/testing-manifests/embed.go:
```diff
@@ -0,0 +1,33 @@
+	e2etestfiles "k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed *
```
same question... won't this embed `embed.go`?

liggitt replied:
would it be better to put the embed util in `test/e2e/embed.go` and have it `//go:embed testing-manifests`?

palnabarun replied:
> would it be better to put the embed util in test/e2e/embed.go and have it //go:embed testing-manifests?

I did try this approach of having the embed util in a directory above first. The imports of the util functions were looking weird. For example getting the images and fixtures would need an embed util in `test` directory.

Do think having the embed util for testing-manifests in e2e would sound better while the ones for images and fixtures remain inside those directories?

liggitt replied:
I don't feel that strongly about the location, but I do think we should be more specific about what gets embedded

liggitt commented on test/e2e_node/testing-manifests/embed.go:
```diff
@@ -0,0 +1,33 @@
+	e2etestfiles "k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed *
```
`*.yaml`?

liggitt commented on test/fixtures/embed.go:
```diff
@@ -0,0 +1,33 @@
+	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed *
```
do we care if this embeds `embed.go`?

palnabarun replied:
xref: https://github.com/kubernetes/kubernetes/pull/99829/files#r651720993

liggitt commented on hack/update-translations.sh:
```diff
@@ -71,7 +72,7 @@ if [[ "${generate_pot}" == "true" ]]; then
   kube::util::ensure-temp-dir
   if msgcat -s tmp.pot > "${KUBE_TEMP}/template.pot"; then
-    mv "${KUBE_TEMP}/template.pot" translations/kubectl/template.pot
+    mv "${KUBE_TEMP}/template.pot" "${TRANSLATIONS}/kubectl/template.pot"
```
nit, if this is already under `staging/src/k8s.io/kubectl`, can drop the `kubectl` dir under here

palnabarun replied:
Good catch. Will change that.

palnabarun replied:
On second thought, how about keeping the directory structure as it is now?

I am worried about changing many things at once here. Maybe this optimization can come later on?

liggitt commented on hack/update-translations.sh:
```diff
@@ -81,12 +82,10 @@ fi
 if [[ "${generate_mo}" == "true" ]]; then
   echo "Generating .po and .mo files"
-  for x in translations/*/*/*/*.po; do
+  for x in "${TRANSLATIONS}"/*/*/*/*.po; do
```
if we drop the `kubectl` subdir, would change to:

```suggestion
  for x in "${TRANSLATIONS}"/*/*/*.po; do
```

liggitt commented on test/e2e/e2e_test.go:
```diff
@@ -33,10 +33,14 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework/testfiles"
-	"k8s.io/kubernetes/test/e2e/generated"
+	e2etestingmanifests "k8s.io/kubernetes/test/e2e/testing-manifests"
+	e2enodetestingmanifests "k8s.io/kubernetes/test/e2e_node/testing-manifests"
```
does `test/e2e` depend on `test/e2e_node` today? should it?

palnabarun replied:
`test/e2e` shouldn't depend on `test/e2e_node`. I am dropping that import.

palnabarun replied:
I rechecked the file references in `test/e2e` and couldn't find any references to the manifests in `test/e2e_node/testing-manifests`.

liggitt commented on test/e2e/e2e_test.go:
```diff
@@ -33,10 +33,14 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework/config"
 	"k8s.io/kubernetes/test/e2e/framework/testfiles"
-	"k8s.io/kubernetes/test/e2e/generated"
+	e2etestingmanifests "k8s.io/kubernetes/test/e2e/testing-manifests"
+	e2enodetestingmanifests "k8s.io/kubernetes/test/e2e_node/testing-manifests"
```
are packages with hyphens ok?

palnabarun replied:
I couldn't find any concrete evidence to say otherwise. I didn't want to change how the manifests are stored and the directories.

liggitt replied:
`e2e_node/testing-manifests` makes me hurt a little inside, but agree on minimizing unnecessary movement

palnabarun replied:
> e2e_node/testing-manifests makes me hurt a little inside

Indeed. The mixed directory naming looks weird.

liggitt commented on test/images/embed.go:
```diff
@@ -14,6 +14,20 @@ See the License for the specific language governing permissions and
+	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed *
```
there's a **lot** of stuff under test/images ... were we building it all into generated bindata before? if not, I don't think we should slurp it all into an embedded file like this

palnabarun replied:
Yes. We were generated bindata for all of that (ref: https://github.com/kubernetes/kubernetes/blob/36a19df91baf8d042f8ba13063905c6395737b7e/hack/generate-bindata.sh#L48) barring Markdown, JPG and PNG files.

liggitt replied:
so this expands to include markdown/images/all files? is there a way to avoid that?

palnabarun replied:
Right now, the go:embed syntax doesn't provide a way to exclude files from a filesystem where a wildcard expression includes everything.

We have to specifically include files if we want to pick what we want and what we don't. This is like a gotcha. I am not very happy with this limitation.

BenTheElder replied:
if we can check which files were in use, perhaps inverting to only include what we actually use?

palnabarun replied:
I grep'ed through references of the embedded files and trimmed down the number of files that were being embedded.

ehashman reviewed (commented):
LGTM for `test/e2e_node/*`

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented on test/e2e/testing-manifests/embed.go:
```diff
@@ -0,0 +1,33 @@
+	e2etestfiles "k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed flexvolume guestbook kubectl sample-device-plugin.yaml scheduling/nvidia-driver-installer.yaml statefulset storage-csi
```
looks good... maybe add a README to the testing-manifests dir saying this needs to be updated if new files/folders are added

palnabarun replied:
Sounds good! Added a README at that location.

liggitt commented on test/fixtures/embed.go:
```diff
@@ -0,0 +1,33 @@
+	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+)
+
+//go:embed doc-yaml/user-guide/liveness doc-yaml/user-guide/secrets doc-yaml/user-guide/downward-api doc-yaml/user-guide/update-demo
```
suggest a README.md pointing here in case additional files used by compiled tests are added

palnabarun replied:
Done!

palnabarun committed:
docs: add documentation on adding files to the embedded data

Signed-off-by: Nabarun Pal <pal.nabarun95@gmail.com>

liggitt commented:
/lgtm
/approve

palnabarun commented:
/assign @smarterclayton
(for test/conformance/testdata changes)

dims commented:
/approve
/lgtm

/assign @spiffxp @johnbelamaric

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-conformance-image-test

palnabarun commented:
/test pull-kubernetes-conformance-image-test

spiffxp reviewed (approved):
/approve
/lgtm
